.github/actions/restore-node/action.yml

```
@@ -123,9 +129,33 @@ runs:
         if test -e ~/endo; then
           scripts/get-packed-versions.sh ~/endo | scripts/resolve-versions.sh
         fi
+        cp package.json package.json.orig
         yarn install
+
+        if cmp -s package.json.orig package.json; then
+          # We need to generate a yarn-installed.sum that matches the
+          # unmodified package.json.  As long as we don't explicitly
+          # `yarn install` anywhere other than in this action and in
+          # the bin/agd script, we should be able to reuse even
+          # Endo-overridden built caches successfully.
+          git checkout HEAD -- package.json
+        fi
+        rm package.json.orig
+
+        function findsha1() {
+          find ${1+"$@"} -print0 | xargs -0 sha1sum | sort +1 || true
+        }
+
+        # Find the current list of package.jsons.
+        files=( package.json )
+        while IFS= read -r line; do
+          files+=( "$line" )
+        done < <(yarn -s workspaces info |
+          sed -ne '/"location":/{ s/.*": "//; s!",.*!/package.json!; p; }')
```

This might be read more clearly with information piping strictly left to right
```suggestion
        yarn -s workspaces info |
          sed -ne '/"location":/{ s/.*": "//; s!",.*!/package.json!; p; }' |
        while IFS= read -r line; do
          files+=( "$line" )
        done
```

Is there no other idiom to self-concat the bash array? `while` subshells for every iteration. I don’t expect that’s a problem, but I have a blind spot for bash arrays.

---

Did some fiddling locally and it seems that this should work:
```bash
file+=( $(
  yarn -s workspaces info |
  sed -ne '/"location":/{ s/.*": "//; s!",.*!/package.json!; p; }'
) )
```

That said, I’m weary of bash arrays because it’s hard to predict when they’re going to compose poorly with the command line length limit. File variable concatenation would look like:

```bash
FILES=$(git hash-object -w /dev/null)
FILES=$( git hash-object -w <(
  git cat-file blob "$FILES"
  yarn -s workspaces info |
  sed -ne '/"location":/{ s/.*": "//; s!",.*!/package.json!; p; }'
))
```

---

> `while` subshells for every iteration

That is surprising to me.  Even when `read` is a builtin that has to set a variable in its caller?

---

I use a bash array because I don't trust the escaping (or lack thereof) of other subcommands.  I'd like this script to match `bin/agd`, which doesn't currently require Git.

---

I’ve had trouble with `while` accumulators in the past. Yeah, verified:

```bash
numbers=( 0 )
seq 10 | while read -r number; do
  numbers+="$numbers $number"
done
echo "${numbers[*]}"
# 0
```

---

It's the pipe that starts the subprocess, not the `while`:

```bash
numbers=( 0 )
while read -r number; do
  numbers+="$numbers $number";
done < <(seq 10)
echo "${numbers[*]}"
```

results in:

```
00 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 600 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 6 700 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 600 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 6 7 800 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 600 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 6 700 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 600 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 6 7 8 900 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 600 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 6 700 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 600 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 6 7 800 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 600 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 6 700 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 600 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 500 100 1 200 100 1 2 300 100 1 200 100 1 2 3 400 100 1 200 100 1 2 300 100 1 200 100 1 2 3 4 5 6 7 8 9 10
```

---

Ahah! Thank you. That makes sense and rationalizes the right to left flow!